core/src/databases/database.rs

```diff
@@ -16,9 +40,86 @@ impl Database {
             data_source_id: data_source_id.to_string(),
             database_id: database_id.to_string(),
             name: name.to_string(),
+            db_type: DatabaseType::LOCAL,
+        }
+    }
+
+    pub async fn get_schema(
+        &self,
+        project_id: i64,
+        store: Box<dyn Store + Sync + Send>,
+    ) -> Result<DatabaseSchema> {
+        match self.db_type {
+            DatabaseType::REMOTE => Err(anyhow!("Remote DB not implemented.")),
+            DatabaseType::LOCAL => {
+                let project = Project::new_from_id(project_id);
+                let (tables, _) = store
+                    .list_databases_tables(&project, &self.data_source_id, &self.database_id, None)
+                    .await?;
+                let (rows, _) = store
+                    .list_database_rows(
+                        &project,
+                        &self.data_source_id,
+                        &self.database_id,
+                        &None, // get rows from all tables
+                        None,
+                    )
+                    .await?;
+
+                // group rows by table_id
+                let mut table_rows: HashMap<String, Vec<DatabaseRow>> = HashMap::new();
+
+                for row in rows {
+                    let table_id = row.table_id();
+                    if !table_rows.contains_key(table_id) {
+                        table_rows.insert(table_id.to_string(), Vec::new());
+                    }
+                    table_rows.get_mut(table_id).unwrap().push(row);
+                }
+
+                let table_by_id = tables
+                    .iter()
+                    .map(|table| (table.table_id().to_string(), table))
+                    .collect::<HashMap<String, &DatabaseTable>>();
+
+                let table_ids_with_rows = table_rows.keys().collect::<Vec<&String>>();
+
+                let schema_tables = table_ids_with_rows
+                    .par_iter()
+                    .map(|table_id| {
+                        let table = *table_by_id.get(*table_id).unwrap();
+                        let rows = table_rows.get(*table_id).unwrap();
+
+                        let row_contents =
+                            (rows.iter().map(|x| x.content())).collect::<Vec<&Value>>();
+                        let table_schema = TableSchema::from_rows(&row_contents)?;
```

Comment:
I thought we said we would store this in DB / update or check on upsert?

Reply:
I thought we'd implement the optimizations such as pre-building and caching schema at the end. That way we'll get to start playing with this thing faster. Do you think it's important that we optimize now ?

Reply:
Fine with it. But that means that the DatabaseSchema block will be slow and moving to it a bit painful. But def fine to move forward without caching.

Reply:
caching stuff can very well be on critical path before we release, but not required to start iterating on it

Reply:
yes